File: FastMines_android/src/main/java/fmg/android/app/model/items/MosaicGroupDataItem.java
package fmg.android.app.model.items;

import androidx.databinding.Bindable;

import java.util.concurrent.ThreadLocalRandom;

import fmg.android.app.BR;
import fmg.android.img.AndroidBitmapView;
import fmg.android.img.MosaicGroupImg;
import fmg.common.geom.BoundDouble;
import fmg.core.img.MosaicGroupModel;
import fmg.core.types.EMosaicGroup;

import static fmg.core.img.PropertyConst.PROPERTY_MODEL_BURGER_MENU_PADDING;
import static fmg.core.img.PropertyConst.PROPERTY_MOSAIC_GROUP;

/** Mosaic group item for data model */
public class MosaicGroupDataItem extends BaseDataItem<EMosaicGroup, MosaicGroupModel, AndroidBitmapView<MosaicGroupModel>, MosaicGroupImg.MosaicGroupAndroidBitmapController> {

    public MosaicGroupDataItem(EMosaicGroup eMosaicGroup) {
        super(eMosaicGroup);
        setTitle(eMosaicGroup==null ? null : eMosaicGroup.getDescription());
    }

    @Bindable
    public EMosaicGroup getMosaicGroup()                          { return getUniqueId(); }
    public void         setMosaicGroup(EMosaicGroup eMosaicGroup) {        setUniqueId(eMosaicGroup); }

    @Deprecated
    public String getUnicodeChar() { return getMosaicGroup() == null ? null : Character.toString(getMosaicGroup().unicodeChar(false)); }

    @Override
    public MosaicGroupImg.MosaicGroupAndroidBitmapController getController() {
        if (this.controller == null) {
            var ctrl = new MosaicGroupImg.MosaicGroupAndroidBitmapController(getUniqueId());
            var m = ctrl.getModel();
            m.setBorderWidth(3);
            m.setRotateAngle(ThreadLocalRandom.current().nextInt(90));
            setController(ctrl);
        }
        return this.controller;
    }

    @Bindable
    public BoundDouble getPaddingBurgerMenu() {
        BoundDouble pad = getController().getBurgerMenuModel().getPadding();
        double zoom = getZoom();
        return new BoundDouble(pad.left / zoom, pad.top / zoom, pad.right / zoom, pad.bottom / zoom);
    }

    public void setPaddingBurgerMenu(BoundDouble paddingBurgerMenu) {
        getController().getBurgerMenuModel().setPadding(zoomPadding(paddingBurgerMenu));
    }

    @Bindable
    public String getContentDescription() {
        return "MosaicGroupItem:" + getMosaicGroup().getDescription();
    }

    @Override
    protected void firePropertyChanged(String propertyName) {
        if (isDisposed())
            return;

        super.firePropertyChanged(propertyName);

        switch (propertyName) {
        case PROPERTY_UNIQUE_ID:
            getController().getModel().setMosaicGroup(getMosaicGroup());
            break;
        case PROPERTY_MOSAIC_GROUP:
            super.firePropertyChanged(PROPERTY_UNIQUE_ID);
            break;
        }
    }

    @Override
    protected void firePropertyChangedAsync(String propertyName) {
        if (isDisposed())
            return;

        super.firePropertyChangedAsync(propertyName);

        // refire as android data binding event
        switch (propertyName) {
        case PROPERTY_MOSAIC_GROUP             : notifyPropertyChanged(BR.mosaicGroup      ); break;
        case PROPERTY_MODEL_BURGER_MENU_PADDING: notifyPropertyChanged(BR.paddingBurgerMenu); break;
        }
    }

    @Override
    public void close() {
        getController().getBurgerMenuModel().setListener(null);
        super.close();
    }

}

File: FastMines_uwp/src/fmg/uwp/app/Model/Items/MosaicGroupDataItem.cs
using System.ComponentModel;
using Microsoft.Graphics.Canvas;
using Fmg.Common;
using Fmg.Common.Notifier;
using Fmg.Common.Geom;
using Fmg.Core.Types;
using Fmg.Core.Img;
using MosaicGroupView       = Fmg.Uwp.Img.Win2d.MosaicGroupImg.CanvasBmpView;
using MosaicGroupController = Fmg.Uwp.Img.Win2d.MosaicGroupImg.CanvasBmpController;

namespace Fmg.Uwp.App.Model.Items {

    /// <summary> Mosaic group item for data model </summary>
    public class MosaicGroupDataItem : BaseDataItem<EMosaicGroup?, MosaicGroupModel, MosaicGroupView, MosaicGroupController> {

        public MosaicGroupDataItem(EMosaicGroup? eMosaicGroup)
            : base(eMosaicGroup)
        {
            Title = eMosaicGroup?.GetDescription();
        }

        public EMosaicGroup? MosaicGroup {
            get => UniqueId;
            set { UniqueId = value; }
        }

        public override MosaicGroupController Entity {
            get {
                if (entity == null) {
                    var tmp = new MosaicGroupController(MosaicGroup, CanvasDevice.GetSharedDevice());
                    var m = tmp.Model;
                    m.BorderWidth = 3;
                    m.RotateAngle = ThreadLocalRandom.Current.Next(90);
                    tmp.BurgerMenuModel.PropertyChanged += OnBurgerMenuModelPropertyChanged;
                    Entity = tmp; // call this setter
                }
                return entity;
            }
        }

        public BoundDouble PaddingBurgerMenu {
            get {
                var pad = Entity.BurgerMenuModel.Padding;
                var zoom = Zoom();
                return new BoundDouble(pad.Left / zoom, pad.Top / zoom, pad.Right / zoom, pad.Bottom / zoom);
            }
            set {
                Entity.BurgerMenuModel.Padding = ZoomPadding(value);
            }
        }

        protected override void OnPropertyChanged(object sender, PropertyChangedEventArgs ev) {
            base.OnPropertyChanged(sender, ev);

            switch(ev.PropertyName) {
            case nameof(this.UniqueId):
                _notifier.FirePropertyChanged(nameof(this.MosaicGroup)); // recall with another property name
                break;
            }
        }

        protected void OnBurgerMenuModelPropertyChanged(object sender, PropertyChangedEventArgs ev) {
            switch (ev.PropertyName) {
            case nameof(BurgerMenuModel.Padding):
                if (ev is PropertyChangedExEventArgs<BoundDouble> evx)
                    _notifier.FirePropertyChanged(ZoomPadding(evx.OldValue), ZoomPadding(evx.NewValue), nameof(this.PaddingBurgerMenu));
                else
                    _notifier.FirePropertyChanged(nameof(this.PaddingBurgerMenu));
                break;
            }
        }

        protected override void Disposing() {
            Entity.BurgerMenuModel.PropertyChanged -= OnBurgerMenuModelPropertyChanged;
            base.Disposing();
        }

    }

}
